keras_cv/layers/feature_pyramid.py

```diff
@@ -0,0 +1,208 @@
+                not isinstance(output_layers, dict)
+                or sorted(output_layers.keys()) != self.pyramid_levels
+            ):
+                raise ValueError(
+                    f"Expect output_layers to be a dict with keys as "
+                    f"{self.pyramid_levels}, got {output_layers}"
+                )
+            self.output_layers = output_layers
+
+        # the same upsampling layer is used for all levels
+        if not top_down_op:
+            self.top_down_op = tf.keras.layers.UpSampling2D(size=2)
+        else:
+            self.top_down_op = top_down_op
+        # the same merge layer is used for all levels
+        if not merge_op:
+            self.merge_op = tf.keras.layers.Add()
+        else:
+            self.merge_op = merge_op
+
+    def call(self, features):
+        # Note that this assertion might not be true for all the subclasses. It is
+        # possible to have FPN that has high levels than the height of backbone outputs.
+        if (
+            not isinstance(features, dict)
+            or sorted(features.keys()) != self.pyramid_levels
+        ):
+            raise ValueError(
+                "Expect the input features to be a dict with int keys that"
+                f"matches to the {self.pyramid_levels}, got: {features}"
+            )
+        return self.build_feature_pyramid(features)
+
+    def build_feature_pyramid(self, input_features):
+        # To illustrate the connection/topology, the basic flow for a FPN with level
+        # 3, 4, 5 is like below:
+        #
+        # input_l5 -> conv2d_1x1_l5 ----V---> conv2d_3x3_l5 -> output_l5
+        #                               V
+        #                          upsample2d
+        #                               V
+        # input_l4 -> conv2d_1x1_l4 -> Add -> conv2d_3x3_l4 -> output_l4
+        #                               V
+        #                          upsample2d
+        #                               V
+        # input_l3 -> conv2d_1x1_l3 -> Add -> conv2d_3x3_l3 -> output_l3
+
+        output_features = {}
+        reversed_levels = list(sorted(input_features.keys(), reverse=True))
+        top_level = reversed_levels[0]
+        for level in reversed_levels:
+            output = self.lateral_layers[level](input_features[level])
+            if level < top_level:
+                # for the top most output, it doesn't need to merge with any upper stream
+                # outputs
+                upstream_output = self.top_down_op(output_features[level + 1])
+                output = self.merge_op([output, upstream_output])
+            output_features[level] = output
+
+        # Post apply the output layers so that we don't leak them to the down stream level
+        for level in reversed_levels:
+            output_features[level] = self.output_layers[level](output_features[level])
+
+        return output_features
+
+    # TODO(scottzhu): Add serialization with get_config/from_config(). It might have
```

Comment:
As long as we save the original inputs in the constructor, I think get_config()/from_config() will actually automatically handle sub-layers as long as sub-layers are registered as serializables themselves